### ISSUE TITLE
HEAD requests are resolving to wildcard routes instead of named routes

### ISSUE THREAD

**Issue by andokai** [fastify/fastify#4866]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.0.0

### Node.js version

20.3.1

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

macOS Ventura 13.4.1

### Description

A HEAD request to a named route is resolving to the wildcard route instead while GET requests resolve correctly.

### Steps to Reproduce

I've put a simple repo here https://github.com/andokai/head-issue

The issue is demonstrated in the failing test in `example.test.js`.

### Expected Behavior

The content-length header for the HEAD request should match the equivalent header for the GET request and it shouldn't match the header for the GET request of the wildcard route.

**Comment by ashutoshdhande** [comment]
Hi there! You may find it helpful to look into issue #2061 as it contains some valuable information, in my opinion. Additionally, it would be great if you could share the responses you receive for both the GET and HEAD requests. could you please include a small relevant code snippet? little cumbersome to browse.

**Comment by ashutoshdhande** [comment]
Hello @andokai,

I have just tested a snippet to handle both `GET` and `HEAD` requests, and they are functioning correctly in my case. It's worth mentioning that we don't need to explicitly set up a separate route for `HEAD` requests since they automatically go through the `GET` route. I used Postman to verify the `content-length` attribute, and it consistently showed the same value of 909 in both cases. You can find the snippet of the headers I received for both requests in the following image:

![image](https://github.com/fastify/fastify/assets/50861363/82e56f19-8a16-4604-8d77-c15d87697b3e)

Additionally, here is the link to my gist for reference: [https://gist.github.com/ashutoshdhande/4895b5ae808ed985c0c821a375425153](url)

If any update please provide.

**Comment by ashutoshdhande** [comment]
@andokai, I performed further tests and discovered that the order of route definition matters. In my previous response, I overlooked testing the presence of a wildcard route. The reason behind this behavior is that routes are scanned from top to bottom. If the wildcard route is positioned above our desired route, such as `/example`, the `HEAD` request is directed to the wildcard route. However, this behavior is not observed with `GET` requests, as they function correctly. I encountered similar behavior when using both Express and Fastify. I will update the gist provided above to reflect these findings.

TL;DR: Set wildcard route at the bottom most in your code (quick fix)

gist containing both fastify and express comparison [https://gist.github.com/ashutoshdhande/4895b5ae808ed985c0c821a375425153](url)

**Comment by andokai** [comment]
@ashutoshdhande Thanks for commenting. What you'll notice from the example project is I have defined my routes using folders, and in that scenario wildcard routes will always be loaded first. The workaround is appreciated however but hopefully this gets fixed at some point.

**Comment by ashutoshdhande** [comment]
Does that `:other` folder refers to the wildcard route?

**Comment by andokai** [comment]
> Does that `:other` folder refers to the wildcard route?

Yes

**Comment by ashutoshdhande** [comment]
@andokai here's my guess wildcard will be used mostly for page not found, so `setNotFoundHandler` can also be a good alternate

**Comment by andokai** [comment]
> @andokai here's my guess wildcard will be used mostly for page not found, so `setNotFoundHandler` can also be a good alternate

I very much doubt it. In my real project's case the wildcard route is the generic route handler so for example you want to retrieve a specific record from a db you'd have `/:table/:id`.

**Comment by metcoder95** [comment]
Did you try without `autoload`?
So far, the automatic creation of the `HEAD` routes handles this scenario.
I'll try to revisit this one with autoload soon

**Comment by andokai** [comment]
Hi @metcoder95, I can workaround the issue but thanks for the suggestion. I was just letting you know of the existence of the issue as I don't see it mentioned anywhere else.

**Comment by metcoder95** [comment]
Thanks for the heads up 🙇 , it seems is mostly an issue with `autoload` then

**Comment by andokai** [comment]
No thanks necessary. I'm just sorry I haven't got the time to attempt to find the cause. 🙁

**Comment by metcoder95** [comment]
No worries at all, feel free to take your time and revisit the issue with different combinations. Let us know if you need any support on finding the root cause 😄

**Comment by metcoder95** [comment]
Closing the issue. Feel free to open a new one @andokai if you are able to dig deeper into it 🙂

**Comment by andokai** [comment]
> Closing the issue. Feel free to open a new one @andokai if you are able to dig deeper into it 🙂

Hi @metcoder95, that's not a problem at all. I do still intend to get back to this once I get out form under my current project.

My only question is one of semantics. Closing it as completed suggest that the issue has been addressed. But it still exists right?!

**Comment by Eomm** [comment]
Printing out the routes:

```js
fastify.addHook('onReady', async function hook () {
    console.log(fastify.printRoutes())
})
```

```
└── / (GET, HEAD)
    ├── example (GET)
    │   └── / (GET)
    └── :other (GET, HEAD)
        └── / (GET, HEAD)
```


I can't see the `head` route in the `example` route.
I think the issue is hidden in some optimizer here or in autoload tho.

**Comment by metcoder95** [comment]
> My only question is one of semantics. Closing it as completed suggest that the issue has been addressed. But it still exists right?!

yes, my bad 👍
But @Eomm reopened so let’s follow that. 

> I think the issue is hidden in some optimizer here or in autoload tho.

My first assumption is within the auto load as if reproducible without it, it might be an issue when  siblings routes 🤔

**Comment by Uzlopak** [comment]
Either we set the routes wrong in find-my-way or it is a bug in find-my-way. We should first test in find-my-way if the bug is reproducable there. If not we know it is an issue here.

**Comment by Eomm** [comment]
Why the HEAD /example/ is missing from the print?
Are we adding that route handler?

**Comment by metcoder95** [comment]
This is the result I've got if I register the routes without `autoload`:
```shell
└── / (GET, HEAD)
    ├── example (GET, HEAD)
    ├── other (GET, HEAD)
    ├── nested/
    │   ├── one (GET, HEAD)
    │   ├── two (GET)
    │   └── :param (GET, HEAD)
    └── another/one (GET, HEAD)
```

It seems that in `nested` context, one of them is missed, but the order matters.
If the parametrised is registered first, the static ones are not registered. But if the static ones are declared first, just one gets registered meanwhile the rest is not, but parametrised ones are still registered.

```js
const fastify = require('fastify')({
  logger: true
})

fastify.get('/example', async (request, reply) => {
  reply.type('application/json').code(200)

  puppet = await puppeteer.launch({
    headless: false
  })

  return { hello: 'world' }
})

fastify.get('/other', async (request, reply) => {
  reply.type('application/json').code(200)

  puppet = await puppeteer.launch({
    headless: false
  })

  return { hello: 'world' }
})

fastify.register(
  (instance, _opts, done) => {
    instance.get('/one', () => {})
    instance.get('/:param', () => {})
    instance.get('/two', () => {})

    done();
  },
  {
    prefix: '/nested'
  }
)


fastify.register(
  (instance, _opts, done) => {
    instance.get('/one', () => {})

    done();
  },
  {
    prefix: '/another'
  }
)

fastify.addHook('onReady', async function hook () {
  console.log(fastify.printRoutes())
})

fastify.listen().then(console.log)

```

**Comment by Rantoledo** [comment]
Hey guys,
Just wanted to say that I've been researching this issue for the last couple of days, and I think I have a solution.
It took me a while because it was my first time trying to contribute to an open-source project and actually writing something for this project.
Hope to open a detailed pull request in the next couple of days and discuss it with you guys.
